parser_library/src/config/assembler_options.cpp

```diff
@@ -116,39 +116,58 @@ void to_json(nlohmann::json& j, const assembler_options& p)
 }
 
 namespace {
-std::optional<bool> get_goff_from_json(const nlohmann::json& j)
-{
-    bool present = false;
-    bool goff = false;
-    bool xobject = false;
+using from_json_helper_function = std::function<void(const nlohmann::json&, assembler_options&)>;
 
-    if (auto it = j.find("GOFF"); it != j.end())
-    {
-        it->get_to(goff);
-        present = true;
-    }
+const auto profile_json = from_json_helper_function(
+    [](const nlohmann::json& val, assembler_options& p) { optional_from_json(val, p.profile); });
+
+const auto sysparm_json = from_json_helper_function(
+    [](const nlohmann::json& val, assembler_options& p) { optional_from_json(val, p.sysparm); });
+
+const auto system_id_json = from_json_helper_function(
+    [](const nlohmann::json& val, assembler_options& p) { optional_from_json(val, p.system_id); });
+
+const auto optable_json = from_json_helper_function(
+    [](const nlohmann::json& val, assembler_options& p) { optional_from_json(val, p.optable); });
+
+const auto goff_json =
+    from_json_helper_function([](const nlohmann::json& val, assembler_options& p) { optional_from_json(val, p.goff); });
 
-    if (auto it = j.find("XOBJECT"); it != j.end())
+struct string_hash
+{
+    using is_transparent = void;
+
+    std::size_t operator()(std::string_view sv) const
     {
-        it->get_to(xobject);
-        present = true;
+        std::hash<std::string_view> hasher;
+        return hasher(sv);
     }
-    if (present)
-        return goff || xobject; // GOFF and XOBJECTS are synonyms
-    else
-        return std::nullopt;
-}
+};
+
+const std::unordered_map<std::string, from_json_helper_function, string_hash, std::equal_to<>> from_json_map = {
+    { "PROFILE", profile_json },
+    { "SYSPARM", sysparm_json },
+    { "SYSTEM_ID", system_id_json },
+    { "GOFF", goff_json },
+    { "XOBJECT", goff_json },
+    { "OPTABLE", optable_json },
+    { "MACHINE", optable_json }
+};
 } // namespace
 
 void from_json(const nlohmann::json& j, assembler_options& p)
 {
     if (!j.is_object())
-        throw nlohmann::json::other_error::create(501, "asm_options must be an object.");
+        throw nlohmann::json::other_error::create(501, "asm_options must be an object.", j);
 
-    p.goff = get_goff_from_json(j);
-    optional_from_json(j, "OPTABLE", p.optable);
-    optional_from_json(j, "PROFILE", p.profile);
-    optional_from_json(j, "SYSPARM", p.sysparm);
-    optional_from_json(j, "SYSTEM_ID", p.system_id);
+    for (const auto& item : j.items())
+    {
+        const auto& key = item.key();
+
+        if (auto it = from_json_map.find(key); it != from_json_map.end())
+        {
+            it->second(item.value(), p);
+        }
+    }
```

Comment:
I think that the original solution with a function call per field was both more compact and more readable.
Especially after the change to optional_from_json there is even no reason for the extra goff handling routine.

Reply:
I also thought about it but the scenario where we are checking the presence of all `assembler_option` members in an almost empty `json` object seemed like a real scenario to me and I didn't want to waste resources on these checks.

If your experience is that user usually provides many assembler options and we will be checking them anyway then I can most certainly change this back.

Reply:
I think that performance is not a concern here, but the previous version seems more easily readable to me.

Reply:
Ok